test/integration/consul-container/test/README.md

```diff
@@ -0,0 +1,97 @@
+
+## Getting Started
+### Prerequisites
+If you wish to run or add new test cases, the following are required:
+- install [Go](https://go.dev/) (the version should match that of our CI config's Go image).
+- install [`golangci-lint`](https://golangci-lint.run/usage/install/) 
+- install [`Makefile`](https://www.gnu.org/software/make/manual/make.html)
+- [`Docker`](https://docs.docker.com/get-docker/) required to run tests locally  
+
+### Running Upgrade integration tests
+To run upgrade integration tests locally, here are the commands needed:
+```sh
+make dev-docker
+cd ./test/integration/consul-container
+go test -v -timeout 30m -run ^TestACL_Upgrade_Node_Token$ ./.../upgrade/ --target-image consul --target-version local --latest-image consul --latest-version 1.15 -follow-log=false
+```
+
+The command above runs a single test. To run the entire upgrade test suite, the following command can be used.
+```sh
+go test -v -timeout 30m ./.../upgrade --target-image consul --target-version local --latest-image consul --latest-version 1.15 -follow-log=false
+```
+
+| Tags / Flags      | Default value | Description |
+| -----------         | ----------- | ----------- |
+| --latest-image      | consul    | Refers to the docker image name to be deployed before upgrade
+| --latest-version    | 1.15      | Refers to the version or tag of latest image to be deployed before upgrade
+| --target-image      | consul    | Refers to the image name that will be upgrade to
+| --target-version    | local     | Refers to the version of target image to upgrade to
+| -follow-log         |  false    | This optional flag emit event logs
+
+
+
+## Adding a new upgrade integration test  
+Upgrade integration tests are defined in the [test/integration/consul-container/test/upgrade](/test/integration/consul-container/test/upgrade) subdirectory and new upgrade integration tests should always be added to this location. The test framework uses
+[functional table-driven tests in Go](https://yourbasic.org/golang/table-driven-unit-test/) and using function types to modify the base value for each test case.
+
+For tests with multiple test cases, it should always start by invoking 
+```go
+	type testcase struct {
+		name string
+		create func()
+		extraAssertion func()
+	}
+```
+see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/l7_traffic_management/resolver_default_subset_test.go). For upgrade tests with a single test case, they can be written like
+```go
+	run := func(t *testing.T, oldVersion, targetVersion string) {
+        // insert test
+    }
+	t.Run(fmt.Sprintf("Upgrade from %s to %s", utils.LatestVersion, utils.TargetVersion),
+		func(t *testing.T) {
+			run(t, utils.LatestVersion, utils.TargetVersion)
+		})
+```
+see example [here](./hashicorp/consul/test/integration/consul-container/test/upgrade/acl_node_test.go)
+
+### How it works
+![Upgrade Tests Workflow](util/upgrade_tests_workflow.png?raw=true)
+
```

Comment:
I don't think this 2 paragraph reveals the details of how it works. Let's think about how to rephrase them.

Reply:
Okay, I'm open to ideas on how to phrase it.